File: py/day07.py
# %%
import numpy as np

# %%
if __name__ == "__main__":
    h = np.loadtxt("../data/day07.txt", delimiter=',', dtype=int)
    # h = np.array([16, 1, 2, 0, 4, 2, 7, 1, 2, 14])  # uncomment for testing
    
    part1 = np.abs(h - int(np.median(h))).sum()  # analytical solution
    # part1 = np.min([np.abs(h - i).sum() for i in range(max(h) + 1)])

    # part2 
    part2 = np.min([((h - i) ** 2 + np.abs(h - i)).sum() for i in range(max(h) + 1)]) // 2

    print("Part 1 —", part1)
    print("Part 2 —", part2)

File: rs/day07.rs
use std::cmp;

fn cost_part1(h: &Vec<isize>, pos: isize) -> usize {
    h.iter().map(|x| (*x - pos).abs() as usize).sum()
}

fn cost_part2(h: &Vec<isize>, pos: isize) -> usize {
    h.iter()
    .map(|x| (x - pos).pow(2) + (*x - pos).abs())
    .map(|c| (c / 2) as usize)
    .sum()
}

fn main() {
    let h: Vec<isize> = include_str!("../data/day07.txt")
        .trim().split(',')
        .map(|x| x.parse().unwrap()).collect();
    let n = h.len() as isize;
    

    let part1 = (1..n).map(|d| cost_part1(&h, d)).min().unwrap();
    // let part2 = (1..n).map(|d| cost_part2(&h, d)).min().unwrap();
    
    // analytical solutions
    let mean = h.iter().sum::<isize>() / n;
    let part2 = cmp::min(cost_part2(&h, mean), cost_part2(&h, mean + 1));

    println!("Part 1 — {:?}", part1);
    println!("Part 2 — {:?}", part2);
}